### PR TITLE
[Classical Shadow] Implementation of classical shadows in mitiq

### LINKED ISSUE #1878: Add sampling functions for classical shadows
As a jumping off point to implement Robust Shadow Estimation [[RFC](https://docs.google.com/document/d/1B5FnqQDvoRYap5fGPqzcbp-RXIrUFjbBcLiWIUrLmuA/)], we need a workflow to perform random Pauli measurements on a specified circuit. The suggested workflow is to have two functions (WIP code available at https://github.com/unitaryfund/mitiq-internal/pull/64). One will generate the necessary Pauli strings, and the other will generate the circuits from the supplied strings. Question for Min: will these be public functions? Do we expect users to use them, or will they be used primarily internally?

```py
def generate_pauli_strings(num_qubits: int, num_strings: int) -> List[str]:
    # return "XXXYZYZX" where len string == num_qubits
    ...


def get_rotated_circuits(
    circuit: cirq.Circuit, pauli_strings: List[str]
) -> List[cirq.Circuit]:
    ...
```
Function/variable names should be changed where fit.

To get started, a new directory will need to be added to `mitiq/mitiq` and should have a directory structure similar to the other modules (`zne`, `pec`, etc.). E.g.
```
mitiq/EXAMPLE
├── EXAMPLE.py
├── EXAMPLE_utils.py
├── __init__.py
└── tests
    └── test_EXAMPLE.py
```

cc @min-li

### PR DESCRIPTION
**Pull request by Min-Li**
## Description

---
This is the prototype of classical shadow implementation in Mitiq, w/o calibration.  In this pull request:
 
    - `mitiq/shadows/__init__.py`  initiation file
    - `mitiq/shadows/classical_postprocessing.py` classical shadow postprocessing functions.
    - `mitiq/shadows/quantum_processing.py` perform z-basis measurements.
    - `mitiq/shadows/shadows.py`  a main function of classical shadow has been written.
    - `mitiq/shadows/shadows_example.ipynb` example notebook for this module.
    - `mitiq/shadows/shadows_utils.py` utility functions, including bounds, fidelity, etc.
    - `mitiq/shadows/test/` including test functions for each file, labeled by file names.
---

### License

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.
- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [ ] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [ ] Added myself / the copyright holder to the AUTHORS file

### CODE REVIEW

File: mitiq/shadows/computational_basis_measurement.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    return rotated_circuits
+
+# Stage 1 of Classical Shadows: Measurements
+def shadow_measure_with_executor(
+        circuit: cirq.Circuit,
+        executor: Executor, 
+        n_total_measurements: int, ) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    
+    Given a circuit, calculate the classical shadow of the circuit of aa bit
```
Comment (nathanshammah) [inline mitiq/shadows/computational_basis_measurement.py:56]:
```suggestion
    Given a circuit, calculate the classical shadow of the circuit of a bit
```

File: mitiq/shadows/shadows.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        circuit: The circuit to execute.
+        executor: The executor to use for running the circuit.
+        observables: The observables to measure. If None, the state will be reconstructed.
+        state_reconstruction: Whether to reconstruct the state or estimate the expectation value of the observables.
+        K1: Number of groups of "median of means" used for calibration in rshadow
+        R1: Total number of shots for calibration in rshadow
+        N1: Number of shots per group of "median of means" used for calibration in rshadow N1=R1/K1
+        K2: Number of groups of "median of means" used for shadow estimation
+        R2: Number of shots per group of "median of means" used for shadow estimation
+        N2: Total number of shots for shadow estimation N2=R2/K2
```
Comment (nathanshammah) [inline mitiq/shadows/shadows.py:74]: Use more "pythonic" names. For example, `R1`-->`shots` or `estimation_shots` or `robust_shadow_shots`.

File: mitiq/shadows/shadows_utils.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        Scalar corresponding to the norm.
+    """
+    return float(np.sqrt(np.trace(R.conjugate().transpose() @ R)).reshape(-1).real)
+
+def fidelity(state_vector: np.ndarray, rho: np.ndarray) -> float:
+    """
+    Calculate the fidelity $$F(\rho,\sigma)=\mathrm{Tr}\sqrt{\rho^{1/2}\sigma\rho^{1/2}}$$, when $$\rho=|v\rangle\langle v|$$ is a pure state $$F(\rho,\sigma)=\langle v|\sigma|v\rangle$$.
+
+    Args:
+        v (array): The vector whose norm we want to calculate.
```
Comment (nathanshammah) [inline mitiq/shadows/shadows_utils.py:65]:
```suggestion
        state_vector (array): The vector whose norm we want to calculate.
```

File: mitiq/shadows/shadows_utils.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        num_qubits (int): The number of qubits in the system.
+    
+    Returns:
+        An integer that gives the number of samples required to satisfy the shadow bound.
+    """
+    return int(34* (4**num_qubits) * epsilon**(-2))
+
+
+# based on the theorem, we calculate N,K for the shadow bound
+def shadow_bound_theorem(error: float, observables: list, failure_rate: float)->Tuple[int,int]:
```
Comment (nathanshammah) [inline mitiq/shadows/shadows_utils.py:20]: I would rename this function to match more closely the output (something like "calculate shadow bound")

File: mitiq/shadows/executor_functions.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        measurements = {}
+        for key, value in result.measurements.items():
+            measurements[int(key[2:-1])] = value[0][0]
+        measurements = dict(sorted(measurements.items()))
+        bitstrings = [list(measurements.values())]
+        qubit_indices = tuple(measurements.keys())
+        outcomes.append(MeasurementResult(bitstrings, qubit_indices))
+    return outcomes
+# qiskit simulator
+def qiskit_simulator_shadow_executor_fn(cirq_circuits: List[cirq.Circuit], backend:Any='aer_simulator') -> List[MeasurementResult]:
```
Comment (nathanshammah) [inline mitiq/shadows/executor_functions.py:27]: Missing docstrings. Not sure this function should live in the codebase though.

File: mitiq/shadows/executor_functions.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+from typing import List, Tuple, Dict, Any, Optional, Union
+import qiskit
+from qiskit_aer import Aer
+from mitiq.interface.mitiq_qiskit.conversions import to_qiskit
+import numpy as np
+from tqdm.auto import tqdm, trange
+
+
+# cirq simulator
+def cirq_simulator_shadow_executor_fn(circuits: List[cirq.Circuit], max_batch_size:int=10000) -> List[MeasurementResult]:
```
Comment (nathanshammah) [inline mitiq/shadows/executor_functions.py:13]: Missing docstrings. Not sure this function should live in the codebase though.

**Review by nathanshammah — CHANGES_REQUESTED** [review]
Excited to see this pull request, @Min-Li, the beginning of classical shadows techniques in Mitiq!

Some comments:
- The `shadows_example.ipynb` should not be in the codebase at `mitiq/mitiq`. However it is a nice tutorial, I would include it into the documentation as one of the examples (under `mitiq/docs/source/examples/` and it needs to be formatted like the other examples (1 add the tutorial as a `.md` file; 2. add the link to the `examples` index file; 3. save a thumbnail and add a link to it to the `conf.py` file).
- the `validate` tests in CI are currently failing for some mypy issue. You can find the log [here](https://github.com/unitaryfund/mitiq/actions/runs/5483183553/jobs/9989244387). 
-   Please add the list of the new modules to the API doc file ([`apidoc.md`](https://github.com/unitaryfund/mitiq/blob/master/docs/source/apidoc.md)) in the documentation. Once we have those, you can test locally with `make docs` that they generate correctly on the API-doc page. On remote, you can check the documentation for this PR [here](https://mitiq--1899.org.readthedocs.build/en/1899/).
- In some instances, use more "pythonic" names. For example, `R1`-->`shots` or `estimation_shots` or `robust_shadow_shots`. More context [here](https://peps.python.org/pep-0008/#prescriptive-naming-conventions). Also, I am not sure that the name of the module "mitiq/shadows/computational_basis_measurement.py" is the ideal one 
- Add tests. This is very important

Overall comment: Consider scoping down this PR to address all and only the points mentioned here #1878. You can  even remove some of the functions / modules present in this PR, and add them later. This can speed up the review process.

File: mitiq/shadows/computational_basis_measurement.py (hunk 1 above)
Comment (Min-Li) [reply mitiq/shadows/computational_basis_measurement.py]: thanks, have rewrite

File: mitiq/shadows/shadows.py (hunk 2 above)
Comment (Min-Li) [reply mitiq/shadows/shadows.py]: Thanks, have renamed

File: mitiq/shadows/shadows_utils.py (hunk 3 above)
Comment (Min-Li) [reply mitiq/shadows/shadows_utils.py]: Thanks, have changed

File: mitiq/shadows/shadows_utils.py (hunk 4 above)
Comment (Min-Li) [reply mitiq/shadows/shadows_utils.py]: thanks, have renamed

File: mitiq/shadows/executor_functions.py (hunk 5 above)
Comment (Min-Li) [reply mitiq/shadows/executor_functions.py]: Yes, it would work

Comment (Min-Li) [reply mitiq/shadows/executor_functions.py]: yes, it would work

File: mitiq/shadows/executor_functions.py (hunk 6 above)
Comment (Min-Li) [reply mitiq/shadows/executor_functions.py]: yes, it would work

**Comment by natestemen** [comment]
@Min-Li do we need to keep this PR up? I think the PRs you've opened cover everything here, is that right?

**Comment by Min-Li** [comment]
moved to #1921

**Pull request closed without merging**